javalib/src/main/scala/java/util/ArrayDeque.scala

```diff
@@ -171,55 +193,175 @@ class ArrayDeque[E] private (private val inner: js.Array[E])
 
   def pop(): E = removeFirst()
 
-  def size(): Int = inner.length
+  def size(): Int = {
+    if (isEmpty()) {
+      0
+    } else {
+      val capacity = inner.length // local copy
+      endIndex - startIndex + capacity % capacity
+    }
+  }
 
-  private def failFastIterator(startIndex: Int, nex: (Int) => Int) = {
+  private def failFastIterator(forward: Boolean) = {
```

Comment:
Perhaps, at this point, it would be easier to reason about this if we had separate implementations for forward and backward? The only meaningful logic left in common is the `status` checks.